### ISSUE TITLE
Why samples_per_gpu is set to 1?

### ISSUE THREAD

**Issue by buaazeus** [OpenDriveLab/UniAD#150]
Can it be samples_per_gpu = 2?
Thank you.

**Comment by deepak-vij** [comment]
@YTEP-ZHI , I was also exploring how to increase the batch size. When I tried it with "samples_per_gpu = 2", it failed due to "Assert" error. Based on the Gemini search, i also tried the following workaround. Training job ran fine but there were no performance benefits (better GPU utilization numbers). Changing batch size, is that a much bigger effort or something you folk have not got around to it yet? Just curious. Thanks.
++++++++++++++++++++++++++++++++++++++
optimizer = dict(
    type='AdamW',
    lr=4e-4,  # Doubled for effective batch size 2
    paramwise_cfg=dict(
        custom_keys={
            'img_backbone': dict(lr_mult=0.1),
        }),
    weight_decay=0.01)

optimizer_config = dict(
    type='GradientCumulativeOptimizerHook', # Explicitly define the hook type
    cumulative_iters=2,                      # The accumulation count
    grad_clip=dict(max_norm=35, norm_type=2)
)

**Comment by deepak-vij** [comment]
This is what Gemini's explanation is why samples_per_gpus=2 result into assert error and the work around is using "he "Gradient Accumulation" Trick as mentioned previously. Although, I tried that, it did not help either. GPU utilization still stay low for me.
  
This specific AssertionError in UniAD 2.0 occurs because the current implementation of UniAD (and many temporal-based MMDetection3D models) does not support a batch size larger than 1 per GPU. This is a structural limitation of how the temporal queues and BEV (Bird's Eye View) features are managed in the code. The model expects a single sequence of frames per GPU to maintain the temporal history correctly. UniAD uses recurrent temporal fusing.

Frame 1 is processed and its BEV features are saved to a "hidden state" (queue).

Frame 2 is then processed, looking back at the "hidden state" of Frame 1.

If you have a batch size of 2, the GPU receives [Sequence A, Frame 1] and [Sequence B, Frame 1] at the same time. The current code is not indexed to separate the "hidden states" of two different cars/sequences, so they would "leak" into each other.
